astropy/coordinates/_optimizations.pyx

```diff
@@ -0,0 +1,83 @@
+# cython: language_level = 3
+import numpy as np
+cimport numpy as np
+cimport cython
+from libc.math cimport isfinite as _isfinite, M_PI
+from libc.stdint cimport int32_t, int64_t
+
+
+ctypedef fused dtype_t:
+    int32_t
+    int64_t
+    float
+    double
+
+
+cdef bint isfinite(dtype_t val):
+    if dtype_t is int32_t or dtype_t is int64_t:
+        return True
+    else:
+        return _isfinite(val)
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.nonecheck(False)
+@cython.cdivision(True)
+cpdef _wrap_at(np.ndarray[dtype_t] angle, dtype_t wrap_angle, double full_circle):
```

Comment:
In case you stay with cython, it's better to use memory views to pass numpy arrays (https://cython.readthedocs.io/en/latest/src/userguide/memoryviews.html): 
```suggestion
def _wrap_at(dtype_t[:] angle, dtype_t wrap_angle, double full_circle):
```
(also the function is used only from Python, so no need for `cpdef`).

Reply:
> In case you stay with cython, it's better to use memory views to pass numpy arrays

That was a left over from when I tried to have the nd loop in cython. Will change.

Reply:
@saimn It seems read-only memory-views are not supported for fused types, so I could only switch for `wrap_angle` not for `needs_wrapping`.